core/src/main/scala/spire/optional/Perm.scala

```diff
@@ -0,0 +1,120 @@
+      .mkString("Perm(", ", ", ")")
+  }
+
+  private lazy val inverseMapping = mapping.map(_.swap)
+
+  /** The preimage of `k` (so that `apply(invert(k)) == k`). */
+  def invert(k: Int): Int = inverseMapping.getOrElse(k, k)
+
+  /** The inverse permutation, which composes with this to yield the identity.
+    * Note that this uses a lazy inverse map, so the first call may be O(n)
+    * rather than O(1).
+    */
+  def inverse: Perm = new Perm(inverseMapping)
+
+  /** The set of points moved by this permutation (equivalent to preimage). */
+  def image: Set[Int] = mapping.keySet
+
+  /** Permute a seq as long as all the moved points are valid indices. */
+  def permute[A, SA <: SeqLike[A, SA]](seq: SA)(implicit cbf: Factory[A, SA]): Opt[SA] = {
+    if (image.isEmpty) return Opt(seq)
+    if (image.max >= seq.size) return Opt.empty[SA]
+    val builder = cbf.newBuilder
+    cforRange(0 until seq.size) { k =>
+      builder += seq(invert(k))
+    }
+    Opt(builder.result())
+  }
+
+  /** Compose this with another `Perm` with this permutation applied last. */
+  def compose(that: Perm): Perm = new Perm(
+    (this.image | that.image)
+      .map(k => k -> this(that(k)))
+      .filter(Function.tupled(_ != _))
+      .toMap
+  )
+
+  /** Compose this with another `Perm` with this permutation applied first. */
+  def andThen(that: Perm): Perm = that.compose(this)
+}
+
+object Perm {
+  /** A `Perm` given a `Map[Int, Int]` from preimage to image. */
+  def apply(mapping: Map[Int, Int]): Perm = {
+    require(mapping.values.toSet == mapping.keySet, "Image and preimage must be the same.")
+    require(mapping.keys.forall(_ >= 0), "Perm indices must be non-negative.")
+    new Perm(mapping.filter(Function.tupled(_ != _)))
+  }
+
+  /** A `Perm` given a collection of preimage/image pairs. */
+  def apply(pairs: (Int, Int)*): Perm = apply(Map(pairs: _*))
+
+  /** A `Perm` constructed by cycling args so each n,,i,, maps to n,,i+1,,.
+    *
+    * [[Perm#apply(n0:Int*]] can be called subsequently to express any `Perm` as
+    * the product of cycles rather than explicit preimage/image pairs. This is
+    * provided as a convenience for constructing permutations. eg.,
+    * `Perm(1,3)(2,4)` is shorthand for `Perm(1 -> 3, 2 -> 4, 3 -> 1, 4 ->
+    * 2)`. At least two args are required in each parameter list.
+    */
+  def apply(n0: Int, n1: Int, ns: Int*): Perm = {
+    apply((n0 +: n1 +: ns).zip(n1 +: ns :+ n0).toMap)
+  }
+
+  implicit val PermEq: Eq[Perm] = new Eq[Perm] {
+    def eqv(p: Perm, r: Perm): Boolean = p.mapping == r.mapping
+  }
+  implicit val PermIntAction: Action[Int, Perm] = new PermIntAction
+  implicit val PermGroup: Group[Perm] = new PermGroup
+  implicit def PermSeqPartialAction[A, CC[A] <: SeqLike[A, CC[A]]](implicit cbf: Factory[A, CC[A]]): PartialAction[CC[A], Perm] = new PermSeqPartialAction[A, CC[A]]
+}
+
+final class PermIntAction extends Action[Int, Perm] {
+  def actl(perm: Perm, k: Int): Int = perm(k)
```

Comment:
I've made `perm.apply` the left action. Formerly, it was the right action. This is more consistent with our syntax for apply, which has the function name on the left.

This also makes our `Group[Perm].combine` consistent with the `MonoidK[Endo]` instance for Function1, which uses `compose` to combine. Without this change, confusing discrepancies could arise.